clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -677,73 +677,83 @@ constructKernelName(Sema &S, FunctionDecl *KernelCallerFunc,
+  VisitRecordFields(Owner, handlers...);
+}
+
 // Parent contains the FieldDecl or CXXBaseSpecifier that was used to enter
 // the Wrapper structure that we're currently visiting. Owner is the parent type
 // (which doesn't exist in cases where it is a FieldDecl in the 'root'), and
 // Wrapper is the current struct being unwrapped.
 template <typename ParentTy, typename... Handlers>
-static void VisitAccessorWrapper(CXXRecordDecl *Owner, ParentTy &Parent,
-                                 CXXRecordDecl *Wrapper,
-                                 Handlers &... handlers) {
+static void VisitRecord(CXXRecordDecl *Owner, ParentTy &Parent,
+                        CXXRecordDecl *Wrapper, Handlers &... handlers) {
   (void)std::initializer_list<int>{(handlers.enterStruct(Owner, Parent), 0)...};
-  VisitAccessorWrapperHelper(Wrapper, Wrapper->bases(), handlers...);
-  VisitAccessorWrapperHelper(Wrapper, Wrapper->fields(), handlers...);
+  VisitRecordHelper(Wrapper, Wrapper->bases(), handlers...);
+  VisitRecordHelper(Wrapper, Wrapper->fields(), handlers...);
   (void)std::initializer_list<int>{(handlers.leaveStruct(Owner, Parent), 0)...};
 }
 
+int getFieldNumber(const CXXRecordDecl *BaseDecl) {
+  int Members = 0;
+  for (const auto *Field : BaseDecl->fields())
+    ++Members;
+
+  return Members;
+}
+
+template <typename... Handlers>
+static void VisitFunctorBases(CXXRecordDecl *KernelFunctor,
+                              Handlers &... handlers) {
+  VisitRecordHelper(KernelFunctor, KernelFunctor->bases(), handlers...);
+}
+
 // A visitor function that dispatches to functions as defined in
 // SyclKernelFieldHandler for the purposes of kernel generation.
 template <typename... Handlers>
-static void VisitRecordFields(RecordDecl::field_range Fields,
-                              Handlers &... handlers) {
+static void VisitRecordFields(CXXRecordDecl *Owner, Handlers &... handlers) {
 #define KF_FOR_EACH(FUNC)                                                      \
   (void)std::initializer_list<int> { (handlers.FUNC(Field, FieldTy), 0)... }
 
-  for (const auto &Field : Fields) {
+  for (const auto &Field : Owner->fields()) {
     QualType FieldTy = Field->getType();
 
     if (Util::isSyclAccessorType(FieldTy))
+      // FIXME: Does this still work? Check
       KF_FOR_EACH(handleSyclAccessorType);
     else if (Util::isSyclSamplerType(FieldTy))
+      // FIXME: Does this still work? Check
       KF_FOR_EACH(handleSyclSamplerType);
     else if (Util::isSyclSpecConstantType(FieldTy))
+      // FIXME: Does this still work? Check
       KF_FOR_EACH(handleSyclSpecConstantType);
     else if (Util::isSyclStreamType(FieldTy)) {
       // Stream actually wraps accessors, so do recursion
+      // FIXME: Does this still work? Check
       CXXRecordDecl *RD = FieldTy->getAsCXXRecordDecl();
-      VisitAccessorWrapper(nullptr, Field, RD, handlers...);
+      VisitRecord(nullptr, Field, RD, handlers...);
```

Comment:
The 3rd parameter 'RD' seems optimizable to me. I suggest removing it.
It is always depend on the 2nd parameter.
```suggestion
      VisitRecord(nullptr, Field, handlers...);
```

Reply:
Instead, _inside_ VisitRecord func you can write:
CXXRecordDecl *Wrapper = Parent->getType()->getAsCXXRecordDecl();

Reply:
Thank you. I will take a look